src/irr/asset/CGLILoader.cpp

```diff
@@ -46,112 +46,288 @@ namespace irr
+
+			for(std::size_t mipmapLevel = 0; mipmapLevel < texture.levels(); ++mipmapLevel)
+			{
+				glm::tvec3<GLsizei> extent(texture.extent(mipmapLevel));
+
+				ICPUImage::SCreationParams imageInfo;
+				imageInfo.format = format;
+				imageInfo.extent.width = extent.x;
+				imageInfo.extent.height = extent.y;
+				imageInfo.extent.depth = extent.z;
+				imageInfo.mipLevels = texture.levels();
+				imageInfo.arrayLayers = texture.faces();
+				imageInfo.samples = ICPUImage::ESCF_1_BIT;
+				imageInfo.flags = static_cast<IImage::E_CREATE_FLAGS>(0u);
+
+				/*
+					According to https://www.khronos.org/opengl/wiki/Texture_Storage#Anatomy_of_storage,
+					for each mipmap level there is a texture object
+					that may be an array texture or a texture. If we
+					handle an array texture object, the layer variable
+					will be greater than 1. Furthermore if we deal with
+					a cubemap, we have to consider it's faces. So note that
+					array texture being a cubemap has some layers and each
+					layer has faces.
+				*/
+
+				core::smart_refctd_ptr<ICPUImage>** imagesOnAMipMapLevelToHandle;
+				*imagesOnAMipMapLevelToHandle = _IRR_NEW_ARRAY(core::smart_refctd_ptr<ICPUImage>, texture.layers());
+				for(uint16_t i = 0; i < texture.layers(); ++i)
+					imagesOnAMipMapLevelToHandle[i] = _IRR_NEW_ARRAY(core::smart_refctd_ptr<ICPUImage>, texture.faces());
+
+				for (std::size_t layer = 0; layer < texture.layers(); ++layer)					// if it is an array of textures of certain type 
+					for (std::size_t face = 0; face < texture.faces(); ++face)					// if it is a cube, otherwise there is only single face
+					{	
+						// TODO, written as pseudo code
+						// core::smart_refctd_ptr<CCustomAllocatorCPUBuffer<>> buffer = CCustomAllocatorCPUBuffer<>(XSIZEX, XDATAX);	inaccessable!
+						
+						imagesOnAMipMapLevelToHandle[layer][face] = ICPUImage::create(std::move(imageInfo));
+
+						// imagesOnAMipMapLevelToHandle[layer][face]->data = texture.data(layer, face, mipmapLevel);
+						// imagesOnAMipMapLevelToHandle[layer][face]->size = texture.size(layer);
+						// rawImagesBundle = imagesOnAMipMapLevelToHandle[0][0].get();    it is wrong, but we need a way to bind rawImagesBundle and imagesOnAMipMapLevelToHandle in a way iterating through it will be valid
 					}
+			}
 
 			ctx.file = nullptr;
-			core::smart_refctd_ptr<ICPUImage> image = nullptr;
-			image = ICPUImage::create(std::move(imageInfo));
 
-			return SAssetBundle({ image });
+			return SAssetBundle({ core::smart_refctd_ptr<IAsset>(rawImagesBundle, core::dont_grab) });
 		}
 
 		bool CBufferLoaderBIN::isALoadableFileFormat(io::IReadFile* _file) const
 		{
-			return true; // gli provides a function to load files, but we can check files' signature actally if needed
+			return true; // gli provides a function to load files, but we can check files' signature actually if needed
+		}
+
+		inline E_FORMAT getTranslatedGLIFormat(const gli::texture& texture, const gli::gl& glVersion)
```

Comment:
i think you'll need to return a `E_FORMAT,ICPUImageView::SCreationParams::Swizzle` pair

Reply:
Sure, but Swizzles are placed in **SComponentMapping** struct

Reply:
yeah I don't remember all the structs and nested class declarations off by heart ;)